model/task_blocked.go

```diff
@@ -0,0 +1,252 @@
+package model
+
+import (
+	"strings"
+
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/model/build"
+	"github.com/evergreen-ci/evergreen/model/task"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/pkg/errors"
+	"github.com/tychoish/tarjan"
+)
+
+const (
+	taskBlocked  = "blocked"
+	taskPending  = "pending"
+	taskRunnable = "runnable"
+)
+
+// BlockedState returns "blocked," "pending" (unsatisfied dependencies,
+// but unblocked), or "" (runnable) to represent the state of the task
+// with respect to its dependencies
+func BlockedState(t *task.Task, tasksWithDeps []task.Task) (string, error) {
+	if t.DisplayOnly {
+		return blockedStateForDisplayTask(t, tasksWithDeps)
+	}
+	if t.IsPartOfSingleHostTaskGroup() {
+		return blockedStateForTaskGroups(t)
+	}
+	return blockedStatePrivate(t)
+}
+
+func blockedStatePrivate(t *task.Task) (string, error) {
+	if len(t.DependsOn) == 0 {
+		return taskRunnable, nil
+	}
+	dependencyIDs := []string{}
+	for _, d := range t.DependsOn {
+		dependencyIDs = append(dependencyIDs, d.TaskId)
+	}
+	dependentTasks, err := task.Find(task.ByIds(dependencyIDs).WithFields(task.DisplayNameKey, task.StatusKey,
+		task.ActivatedKey, task.BuildVariantKey, task.DetailsKey, task.DependsOnKey))
+	if err != nil {
+		return "", errors.Wrap(err, "error finding dependencies")
+	}
+	taskMap := map[string]*task.Task{}
+	for i := range dependentTasks {
+		taskMap[dependentTasks[i].Id] = &dependentTasks[i]
+	}
+	for _, dependency := range t.DependsOn {
+		depTask := taskMap[dependency.TaskId]
+		state, err := getStateByDependency(depTask, dependency)
+		if err != nil {
+			return "", errors.Wrap(err, "")
+		}
+		if state != taskRunnable {
+			return state, nil
+		}
+	}
+	return taskRunnable, nil
+}
+
+// getStateByDependency determines if the task is still running, is blocked, or has violated the given dependency
+func getStateByDependency(t *task.Task, dependency task.Dependency) (string, error) {
```

Comment:
should we make the state be more "enumish" with a distinct type/etc.

Reply:
I would argue no, this is pretty readable in my opinion, and would be a pretty involved refactor.

Reply:
I also only made this function to simplify blockedStatePrivate, which I thought was a little too involved.

Reply:
my fear is that encoding this information in strings avoids taking advantage of the type system, and while it's pretty clear, it violates the idea that function/method signatures provide the first and best documentation for an API, which I think is important, though certainly it can be limiting, particularly for internal details.

I think this is exacerbated by the fact that the models package is too large (and maybe there's no good way around that,) 

Would it make sense to open a ticket to track the refactoring?

Reply:
I can open another ticket for a refactor.